File: src/foam/dao/UnreliableDAO.java
package foam.dao;

public class UnreliableDAO
  extends ProxyDAO
{
  private double errorRate_ = 0.5;

  public UnreliableDAO(DAO delegate) {
    super();
    setDelegate(delegate);
  }

  public UnreliableDAO(double errorRate, DAO delegate) {
    this(delegate);
    errorRate_ = errorRate;
  }

  @Override
  public foam.core.FObject put_(foam.core.X x, foam.core.FObject obj) {
    if ( Math.random() < errorRate_ ) {
      throw new RuntimeException("UnreliableDAO decided you are unlucky.");
    }

    return super.put_(x, obj);
  }

  @Override
  public foam.core.FObject remove_(foam.core.X x, foam.core.FObject obj) {
    if ( Math.random() < errorRate_ ) {
      throw new RuntimeException("UnreliableDAO decided you are unlucky.");
    }

    return super.remove_(x, obj);
  }

  @Override
  public foam.dao.Sink select_(foam.core.X x, foam.dao.Sink sink, long skip, long limit, foam.mlang.order.Comparator order, foam.mlang.predicate.Predicate predicate) {
    if ( Math.random() < errorRate_ ) {
      throw new RuntimeException("UnreliableDAO decided you are unlucky.");
    }

    return super.select_(x, sink, skip, limit, order, predicate);
  }

  @Override
  public void removeAll_(foam.core.X x, long skip, long limit, foam.mlang.order.Comparator order, foam.mlang.predicate.Predicate predicate) {
    if ( Math.random() < errorRate_ ) {
      throw new RuntimeException("UnreliableDAO decided you are unlucky.");
    }

    super.removeAll_(x, skip, limit, order, predicate);
  }
}

File: src/foam/dao/UnreliableDAO.js
/**
 * @license
 * Copyright 2020 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.dao',
  name: 'UnreliableDAO',
  extends: 'foam.dao.ProxyDAO',

  messages: [
    {
      name: 'CANNOT_PERFORM_ERROR_MSG',
      message: 'UnreliableDAO decided you are unlucky.'
    }
  ],

  properties: [
    {
      class: 'Double',
      name: 'errorRate',
      value: 0.5
    }
  ],

  axioms: [
    {
      name: 'javaExtras',
      buildJavaClass: function(cls) {
        cls.extras.push(
          `
            public UnreliableDAO(DAO delegate) {
              super();
              setDelegate(delegate);
              if ( delegate instanceof ProxyDAO ) {
                setX(((ProxyDAO)delegate).getX());
              }
            }
          
            public UnreliableDAO(double errorRate, DAO delegate) {
              this(delegate);
              setErrorRate(errorRate);
            }
          
            public UnreliableDAO(foam.core.X x, double errorRate, DAO delegate) {
              this(errorRate, delegate);
              setX(x);
            } 
          `
        );
      }
    }
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        if ( Math.random() < getErrorRate() ) {
          throw new RuntimeException(CANNOT_PERFORM_ERROR_MSG);
        }
    
        return super.put_(x, obj);
      `
    },
    {
      name: 'remove_',
      javaCode: `
        if ( Math.random() < getErrorRate() ) {
          throw new RuntimeException(CANNOT_PERFORM_ERROR_MSG);
        }
    
        return super.remove_(x, obj);
      `
    },
    {
      name: 'select_',
      javaCode: `
        if ( Math.random() < getErrorRate() ) {
          throw new RuntimeException(CANNOT_PERFORM_ERROR_MSG);
        }
    
        return super.select_(x, sink, skip, limit, order, predicate);
      `
    },
    {
      name: 'removeAll_',
      javaCode: `
        if ( Math.random() < getErrorRate() ) {
          throw new RuntimeException(CANNOT_PERFORM_ERROR_MSG);
        }
    
        super.removeAll_(x, skip, limit, order, predicate);
      `
    }
  ]
});
